src/pool/srv_pool.c

```diff
@@ -4899,6 +4893,274 @@ ds_pool_svc_check_evict(uuid_t pool_uuid, d_rank_list_t *ranks,
+	if (rc != 0)
+		D_GOTO(out_eventual, rc);
+
+	rc = ABT_eventual_wait(eventual, (void **)&status);
+	if (rc != ABT_SUCCESS)
+		D_GOTO(out_eventual, rc = dss_abterr2der(rc));
+
+	if (*status != 0)
+		D_GOTO(out_eventual, rc = *status);
+
+out_eventual:
+	ABT_eventual_free(&eventual);
+out_bulk:
+	if (bulk != CRT_BULK_NULL)
+		crt_bulk_free(bulk);
+out:
+	return rc;
+}
+
+/**
+ * Send CaRT RPC to pool svc to get list of storage server ranks.
+ *
+ * \param[in]	uuid		UUID of the pool
+ * \param[in]	svc_ranks	Pool service replicas
+ * \param[out]	ranks		Storage server ranks (allocated, caller-freed)
+ *
+ * return	0		Success
+ *
+ */
+int
+ds_pool_svc_ranks_get(uuid_t uuid, d_rank_list_t *svc_ranks,
+		      d_rank_list_t **ranks)
+{
+	int				 rc;
+	struct rsvc_client		 client;
+	crt_endpoint_t			 ep;
+	struct dss_module_info		*info = dss_get_module_info();
+	crt_rpc_t			*rpc;
+	struct pool_ranks_get_in	*in;
+	struct pool_ranks_get_out	*out;
+	uint32_t			 resp_nranks = 2048;
+	d_rank_list_t			*out_ranks = NULL;
+
+	D_DEBUG(DB_MGMT, DF_UUID ": Getting storage ranks\n", DP_UUID(uuid));
+
+	rc = rsvc_client_init(&client, svc_ranks);
+	if (rc != 0)
+		D_GOTO(out, rc);
+
+rechoose:
+	ep.ep_grp = NULL; /* primary group */
+	rc = rsvc_client_choose(&client, &ep);
+	if (rc != 0) {
+		D_ERROR(DF_UUID ": cannot find pool service: " DF_RC "\n",
+			DP_UUID(uuid), DP_RC(rc));
+		goto out_client;
+	}
+
+realloc_resp:
+	rc = pool_req_create(info->dmi_ctx, &ep, POOL_RANKS_GET, &rpc);
+	if (rc != 0) {
+		D_ERROR(DF_UUID ": failed to create POOL_RANKS_GET rpc, "
+			DF_RC "\n", DP_UUID(uuid), DP_RC(rc));
+		D_GOTO(out_client, rc);
+	}
+
+	/* Allocate response buffer */
+	out_ranks = d_rank_list_alloc(resp_nranks);
+	if (out_ranks == NULL)
+		D_GOTO(out_rpc, rc = -DER_NOMEM);
+
+	in = crt_req_get(rpc);
+	uuid_copy(in->prgi_op.pi_uuid, uuid);
+	uuid_clear(in->prgi_op.pi_hdl);
+	in->prgi_nranks = resp_nranks;
+	rc = ranks_get_bulk_create(info->dmi_ctx, &in->prgi_ranks_bulk,
+				   out_ranks->rl_ranks, in->prgi_nranks);
+	if (rc != 0)
+		D_GOTO(out_resp_buf, rc);
+
+	D_DEBUG(DF_DSMS, DF_UUID ": send POOL_RANKS_GET to PS rank %u, "
+		"reply capacity %u\n", DP_UUID(uuid), ep.ep_rank, resp_nranks);
+
+	rc = dss_rpc_send(rpc);
+	out = crt_reply_get(rpc);
+	D_ASSERT(out != NULL);
+
+	rc = rsvc_client_complete_rpc(&client, &ep, rc,
+				      out->prgo_op.po_rc,
+				      &out->prgo_op.po_hint);
+	if (rc == RSVC_CLIENT_RECHOOSE) {
+		/* To simplify logic, destroy bulk hdl and buffer each time */
+		ranks_get_bulk_destroy(in->prgi_ranks_bulk);
+		d_rank_list_free(out_ranks);
+		crt_req_decref(rpc);
+		dss_sleep(1000 /* ms */);
+		D_GOTO(rechoose, rc);
+	}
+
+	rc = out->prgo_op.po_rc;
+	if (rc == -DER_TRUNC) {
+		/* out_ranks too small - realloc with server-provided nranks */
+		resp_nranks = out->prgo_nranks;
+		ranks_get_bulk_destroy(in->prgi_ranks_bulk);
+		d_rank_list_free(out_ranks);
+		crt_req_decref(rpc);
+		dss_sleep(1000 /* ms */);
```

Comment:
Yes I see. This came up in another change that I made a while ago (list containers). I think here, and then, I just used the existing examples in this file as a model (e.g., ds_pool_svc_query). For this PR we should leave it consistent. We can experiment subsequently with removing the sleeps for the rc == -DER_TRUNC scenarios.

For the other cases like rc == RSVC_CLIENT_RECHOOSE it may be helpful because there could be a leader transition in progress and it may be helpful to hold off while a raft election is happening in the pool service.